### PR TITLE
Drkwint/strategies

### PR DESCRIPTION
**Pull request by DrKwint**
I'm only pretty sure this is working, it needs to be run with an eye to the correctness of the gradients.

Ignore the Sacred changes. They won't be shown once #22 is merged.

### CODE REVIEW

File: boosting_strategy.py [hunk 1]
```diff
@@ -64,26 +69,39 @@ def calculate_boosted_losses(boosting_strategy_fn, weak_logits, label_ph, batch_
...
     grads_and_vars = [optimizer.compute_gradients(l) for l in weighted_losses]
     vars = list(zip(*grads_and_vars[0]))[1]
     grads = [[g if g is not None else tf.zeros_like(v) for g, v in gv]
-            for gv in grads_and_vars]
-    grads = [[g * s for g in gs]
-            for gs, s in zip(grads, tf.split(weights_scale, len(grads)))]
+             for gv in grads_and_vars]
+    # scaled_grads = [[
+    #      g * s for g in gs
+    # ] for gs, s in zip(grads, tf.split(weights_scale, len(grads)))]
```
Comment (jakew42) [inline boosting_strategy.py]: Delete commented code.

**Comment by jakew42** [comment]
We should add a way to control number of epochs per classifier in progressive training without having to hardcode.

File: boosting_strategy.py [hunk 2]
```diff
@@ -64,26 +69,39 @@ def calculate_boosted_losses(boosting_strategy_fn, weak_logits, label_ph, batch_
...
     total_grads = [sum(gs) for gs in list(zip(*grads))]
     clipped_grads = [tf.clip_by_value(grad, -1., 1.) for grad in total_grads]
-    final_grads_and_vars = list(zip(clipped_grads, vars))
+    built_grads_and_vars = list(zip(clipped_grads, vars))
+    split_gvs = [[
+        (g * scale, v) for (g, v) in built_grads_and_vars
+        if 'classifier_{}'.format(i) in v.name or 'block_{}'.format(i) in v.name
+    ] for (i, scale) in enumerate(tf.split(weights_scale, len(grads)))]
+    forgotten_gvs = [(g,v) for (g,v) in built_grads_and_vars if 'classifier' not in v.name]
```
Comment (jakew42) [inline boosting_strategy.py]: forgotten_gvs = [(g,v) for (g,v) in built_grads_and_vars if not ('classifier' in v.name or 'block' in v.name)]

**Comment by jakew42** [comment]
Merge approved.
